### PR TITLE
Export BigQuery Table Description to Terra

### PR DESCRIPTION
**Pull request by bfcrampton**
screen shot 2018-10-16 at 5 27 44 pm

### CODE REVIEW

File: api/data_explorer/__main__.py [hunk 1]
```diff
@@ -165,7 +166,21 @@ def _process_bigquery():
...
+                'Unknown format for table name %s. Expected BigQuery project_id.dataset_id.table_name'
+            )
+        project, dataset, table_name = splits
+        client = bigquery.Client(project=project)
+        dataset_ref = client.dataset(dataset, project=project)
+        table_ref = dataset_ref.table(table_name)
+        description = client.get_table(table_ref).description
+        table_names_dict[full_table_name] = description
+
+    app.app.config['TABLE_NAMES'] = table_names_dict
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: TABLES
```
# A dict from table name (project_id.dataset_id.table_name) to table description.
```

File: api/data_explorer/__main__.py [hunk 2]
```diff
@@ -165,7 +166,21 @@ def _process_bigquery():
         sample_file_columns = bigquery_config.get('sample_file_columns', {})
         table_names.sort()
 
-    app.app.config['TABLE_NAMES'] = table_names
+    table_names_dict = OrderedDict()
+    for full_table_name in table_names:
+        splits = full_table_name.split('.')
+        if len(splits) != 3:
+            raise EnvironmentError(
+                'Unknown format for table name %s. Expected BigQuery project_id.dataset_id.table_name'
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: % full_table_name

File: api/data_explorer/__main__.py [hunk 3]
```diff
@@ -165,7 +166,21 @@ def _process_bigquery():
         sample_file_columns = bigquery_config.get('sample_file_columns', {})
         table_names.sort()
 
-    app.app.config['TABLE_NAMES'] = table_names
+    table_names_dict = OrderedDict()
+    for full_table_name in table_names:
+        splits = full_table_name.split('.')
+        if len(splits) != 3:
+            raise EnvironmentError(
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: ValueError

File: api/data_explorer/__main__.py [hunk 4]
```diff
@@ -165,7 +166,21 @@ def _process_bigquery():
...
+                'Unknown format for table name %s. Expected BigQuery project_id.dataset_id.table_name'
+            )
+        project, dataset, table_name = splits
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: By convention: project_id, dataset_id

File: api/data_explorer/__main__.py [hunk 5]
```diff
@@ -165,7 +166,21 @@ def _process_bigquery():
         sample_file_columns = bigquery_config.get('sample_file_columns', {})
         table_names.sort()
 
-    app.app.config['TABLE_NAMES'] = table_names
+    table_names_dict = OrderedDict()
+    for full_table_name in table_names:
+        splits = full_table_name.split('.')
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: (Sorry, I see you copied this code -- I have some comments on it)

Project id can have `.` in it (google.com). So use rsplit to make 3 splits.

File: api/data_explorer/controllers/export_url_controller.py [hunk 6]
```diff
@@ -113,9 +108,10 @@ def _get_entities_dict(cohort_name, query, filter_arr):
             # table_name attribute. Unfortunately RAWLS doesn't allow
             # periods here. RAWLS does allow periods in attributes. So use
             # underscores here and periods in table_name attribute.
-            'name': table_name.replace('.', '_'),
+            'name': name.replace('.', '_'),
             'attributes': {
-                'table_name': table_name
+                'table_name': name,
+                'table_description': description
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: If description isn't set, don't add table_description attribute

File: api/data_explorer/__main__.py (hunk 5 above)
Comment (bfcrampton) [reply api/data_explorer/__main__.py]: Done.

File: api/data_explorer/__main__.py (hunk 3 above)
Comment (bfcrampton) [reply api/data_explorer/__main__.py]: Done.

File: api/data_explorer/__main__.py (hunk 2 above)
Comment (bfcrampton) [reply api/data_explorer/__main__.py]: Oops, done.

File: api/data_explorer/__main__.py (hunk 4 above)
Comment (bfcrampton) [reply api/data_explorer/__main__.py]: Done.

File: api/data_explorer/__main__.py (hunk 1 above)
Comment (bfcrampton) [reply api/data_explorer/__main__.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py (hunk 6 above)
Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

File: api/data_explorer/controllers/export_url_controller.py [hunk 7]
```diff
@@ -113,10 +111,8 @@ def _get_entities_dict(cohort_name, query, filter_arr):
             # table_name attribute. Unfortunately RAWLS doesn't allow
             # periods here. RAWLS does allow periods in attributes. So use
             # underscores here and periods in table_name attribute.
-            'name': table_name.replace('.', '_'),
-            'attributes': {
-                'table_name': table_name
-            }
+            'name': name.replace('.', '_'),
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Do you mind changing this to:
```
'name': name.replace('.', '_').replace(':', '_'),
```

This fixes export for my test project, which has "google.com:" in it. Thanks.

Comment (bfcrampton) [reply api/data_explorer/controllers/export_url_controller.py]: Done.

**Pull request merged**